File: sync-client/packages/gui/src/main/connection-monitor.ts
import { BrowserWindow } from 'electron';
import {
  APIClient,
  CONNECTION_HEALTH_CHECK_INTERVAL_MS,
  CONNECTION_RECONNECT_INTERVAL_MS,
  HEALTH_CHECK_FAILURE_THRESHOLD,
  HEALTH_CHECK_TIMEOUT_MS,
} from '@mporter/core';
import type { ConnectionState } from '@mporter/core';

/** Callback that attempts to reconnect to the server and returns the new connection state. */
type ReconnectCallback = () => Promise<ConnectionState>;

/**
 * Monitors the server connection from the Electron main process.
 *
 * When connected: pings the server periodically. After consecutive failures
 * exceed the threshold, notifies the renderer to switch to offline mode and
 * begins automatic reconnection attempts.
 *
 * When in auto-offline mode: periodically calls the reconnect callback.
 * On success, notifies the renderer to go back online.
 *
 * When the user manually goes offline via "Go Offline": all timers stop.
 * No automatic reconnection is attempted until the user explicitly reconnects.
 */
export class ConnectionMonitor {
  private readonly apiClient: APIClient;
  private mainWindow: BrowserWindow | null = null;
  private reconnectCallback: ReconnectCallback | null = null;
  private healthCheckTimer: ReturnType<typeof setInterval> | null = null;
  private reconnectTimer: ReturnType<typeof setInterval> | null = null;
  private consecutiveFailures = 0;

  constructor(apiClient: APIClient) {
    this.apiClient = apiClient;
  }

  /** Start monitoring with the given window and reconnect callback. */
  start(mainWindow: BrowserWindow, reconnectCallback: ReconnectCallback): void {
    this.mainWindow = mainWindow;
    this.reconnectCallback = reconnectCallback;
  }

  /** Called after a successful connection. Starts health-check polling. */
  notifyConnected(): void {
    this.stopAllTimers();
    this.consecutiveFailures = 0;
    this.startHealthCheck();
  }

  /**
   * Called when the user manually chooses "Go Offline".
   * Stops all timers — no reconnection attempts will be made.
   */
  notifyManualOffline(): void {
    this.stopAllTimers();
    this.consecutiveFailures = 0;
  }

  /**
   * Called on full disconnect (server config cleared).
   * Stops all timers completely.
   */
  notifyDisconnected(): void {
    this.stopAllTimers();
    this.consecutiveFailures = 0;
  }

  /** Stop all monitoring. */
  stop(): void {
    this.stopAllTimers();
  }

  private startHealthCheck(): void {
    this.stopAllTimers();
    this.healthCheckTimer = setInterval(
      () => this.performHealthCheck(),
      CONNECTION_HEALTH_CHECK_INTERVAL_MS,
    );
  }

  private startReconnect(): void {
    this.stopAllTimers();
    this.reconnectTimer = setInterval(
      () => this.attemptReconnect(),
      CONNECTION_RECONNECT_INTERVAL_MS,
    );
  }

  private stopAllTimers(): void {
    if (this.healthCheckTimer !== null) {
      clearInterval(this.healthCheckTimer);
      this.healthCheckTimer = null;
    }
    if (this.reconnectTimer !== null) {
      clearInterval(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }

  private async performHealthCheck(): Promise<void> {
    const ok = await this.apiClient.ping(HEALTH_CHECK_TIMEOUT_MS);
    if (ok) {
      this.consecutiveFailures = 0;
      return;
    }

    this.consecutiveFailures++;
    console.log(
      '[connection-monitor] Health check failed (%d/%d)',
      this.consecutiveFailures,
      HEALTH_CHECK_FAILURE_THRESHOLD,
    );

    if (this.consecutiveFailures >= HEALTH_CHECK_FAILURE_THRESHOLD) {
      console.log('[connection-monitor] Threshold reached — switching to offline mode');
      this.sendToRenderer('connection:statusChange', { offline: true });
      this.startReconnect();
    }
  }

  private async attemptReconnect(): Promise<void> {
    if (!this.reconnectCallback) return;

    try {
      const state = await this.reconnectCallback();
      if (state.connected) {
        console.log('[connection-monitor] Reconnected successfully');
        this.consecutiveFailures = 0;
        this.sendToRenderer('connection:statusChange', { offline: false, connection: state });
        this.startHealthCheck();
      }
    } catch {
      // Reconnect failed — will retry on next interval
    }
  }

  private sendToRenderer(channel: string, data: unknown): void {
    if (this.mainWindow && !this.mainWindow.isDestroyed()) {
      this.mainWindow.webContents.send(channel, data);
    }
  }
}

File: clients/ios/MusicPorter/MusicPorter/Services/ConnectionMonitor.swift
import Foundation
import os

/// Monitors server connection health and manages automatic offline transitions.
///
/// When connected, periodically pings the server. After consecutive failures,
/// transitions to offline mode and begins reconnection attempts.
/// Manual offline (user-initiated) disables auto-reconnect.
@MainActor @Observable
final class ConnectionMonitor {
    // MARK: - Private State

    private weak var appState: AppState?
    @ObservationIgnored private var healthCheckTimer: Timer?
    @ObservationIgnored private var reconnectTimer: Timer?
    @ObservationIgnored private var consecutiveFailures = 0
    @ObservationIgnored private var isManualOffline = false
    @ObservationIgnored private let logger = Logger(subsystem: "com.musicporter", category: "ConnectionMonitor")

    // MARK: - Init

    init(appState: AppState) {
        self.appState = appState
    }

    // MARK: - Lifecycle

    /// Called after a successful connection. Starts health check timer.
    func notifyConnected() {
        stopAllTimers()
        consecutiveFailures = 0
        isManualOffline = false
        startHealthCheckTimer()
        logger.info("Connection monitor started (health check every \(ConnectionConstants.healthCheckIntervalSeconds)s)")
    }

    /// Called when the user manually goes offline (Go Offline button or Use Offline from ReconnectingView).
    /// Stops all timers — no auto-reconnect.
    func notifyManualOffline() {
        stopAllTimers()
        consecutiveFailures = 0
        isManualOffline = true
        logger.info("Manual offline — auto-reconnect disabled")
    }

    /// Called on explicit disconnect. Stops everything.
    func notifyDisconnected() {
        stopAllTimers()
        consecutiveFailures = 0
        isManualOffline = false
        logger.info("Connection monitor stopped (disconnected)")
    }

    // MARK: - Health Check

    private func startHealthCheckTimer() {
        healthCheckTimer?.invalidate()
        healthCheckTimer = Timer.scheduledTimer(
            withTimeInterval: ConnectionConstants.healthCheckIntervalSeconds,
            repeats: true
        ) { [weak self] _ in
            Task { @MainActor [weak self] in
                await self?.performHealthCheck()
            }
        }
    }

    private func performHealthCheck() async {
        guard let appState, appState.apiClient.isConnected else { return }

        let success = await appState.apiClient.ping(
            timeoutSeconds: ConnectionConstants.healthCheckTimeoutSeconds
        )

        if success {
            consecutiveFailures = 0
        } else {
            consecutiveFailures += 1
            logger.warning("Health check failed (\(self.consecutiveFailures)/\(ConnectionConstants.healthCheckFailureThreshold))")

            if consecutiveFailures >= ConnectionConstants.healthCheckFailureThreshold {
                await transitionToOffline()
            }
        }
    }

    // MARK: - Offline Transition

    private func transitionToOffline() async {
        guard let appState else { return }
        logger.warning("Server unreachable — transitioning to offline mode")

        healthCheckTimer?.invalidate()
        healthCheckTimer = nil
        consecutiveFailures = 0

        // Set disconnected state so views stop making server calls
        appState.apiClient.isConnected = false
        appState.backgroundPrefetchService?.stop()

        // Enter offline mode if cache is available
        await appState.checkAndEnterOfflineMode()

        // Start auto-reconnect (not manual offline)
        startReconnectTimer()
    }

    // MARK: - Reconnect

    private func startReconnectTimer() {
        reconnectTimer?.invalidate()
        reconnectTimer = Timer.scheduledTimer(
            withTimeInterval: ConnectionConstants.reconnectIntervalSeconds,
            repeats: true
        ) { [weak self] _ in
            Task { @MainActor [weak self] in
                await self?.attemptReconnect()
            }
        }
        logger.info("Auto-reconnect started (every \(ConnectionConstants.reconnectIntervalSeconds)s)")
    }

    private func attemptReconnect() async {
        guard let appState else { return }
        guard let server = appState.savedServer,
              let apiKey = KeychainService.load() else { return }

        logger.info("Attempting reconnection...")

        do {
            try await appState.connect(server: server, apiKey: apiKey)
            // connect() calls notifyConnected() via AppState wiring,
            // which stops reconnect timer and starts health checks
            appState.isOfflineMode = false
            logger.info("Reconnected successfully")
        } catch {
            logger.info("Reconnection failed: \(error.localizedDescription)")
        }
    }

    // MARK: - Helpers

    private func stopAllTimers() {
        healthCheckTimer?.invalidate()
        healthCheckTimer = nil
        reconnectTimer?.invalidate()
        reconnectTimer = nil
    }
}
